File: src/rag/score.ts
import { METRIC_WEIGHTS } from "./weight";

export interface WeightedMetric {
  score: number | null;
  weight: number;
}

export function calculateOverallScore(
  metrics: WeightedMetric[]
): number {
  let weightedSum = 0;
  let totalWeight = 0;

  for (const metric of metrics) {
    if (metric.score !== null) {
      weightedSum += metric.score * metric.weight;
      totalWeight += metric.weight;
    }
  }

  if (totalWeight === 0) {
    return 0;
  }

  return weightedSum / totalWeight;
}

File: src/rag/score.py
from typing import List, Optional


class WeightedMetric:
    def __init__(self, score: Optional[float], weight: float):
        self.score = score
        self.weight = weight


def calculate_overall_score(metrics: List[WeightedMetric]) -> float:
    """Calculate the weighted overall score from a list of metrics."""
    weighted_sum = 0.0
    total_weight = 0.0

    for metric in metrics:
        if metric.score is not None:
            weighted_sum += metric.score * metric.weight
            total_weight += metric.weight

    if total_weight == 0:
        return 0.0

    return weighted_sum / total_weight
